Framework/PythonInterface/plugins/algorithms/WorkflowAlgorithms/D7AbsoluteCrossSections.py

```diff
@@ -0,0 +1,504 @@
+        entry0 = mtd[sample_ws][0]
+        CreateWorkspace(dataX=entry0.readX(0), dataY=relative_norm_factor, dataE=dataE,
+                        NSpec=entry0.getNumberHistograms(), OutputWorkspace=relative_norm_ws)
+        mtd[relative_norm_ws].getAxis(0).setUnit(entry0.getAxis(0).getUnit().unitID())
+        mtd[relative_norm_ws].getAxis(1).setUnit(entry0.getAxis(1).getUnit().unitID())
+
+        single_efficiency_per_POL = False
+        if mtd[sample_ws].getNumberOfEntries() != mtd[det_efficiency_ws].getNumberOfEntries():
+            single_efficiency_per_POL = True
+        tmp_names = []
+        normalisation_method = self.getPropertyValue('NormalisationMethod')
+        for entry_no, entry in enumerate(mtd[sample_ws]):
+            det_eff_entry_no = int(entry_no / 2)
+            if normalisation_method == 'Vanadium':
+                det_eff_entry_no = 0
+            elif single_efficiency_per_POL:
+                det_eff_entry_no = int(entry_no / 2)
+                if entry_no % 2 != 0:
+                    det_eff_entry_no -= 1
+            ws_name = entry.name() + '_normalised'
+            tmp_names.append(ws_name)
+            Multiply(LHSWorkspace=mtd[det_efficiency_ws][det_eff_entry_no],
+                     RHSWorkspace=relative_norm_ws, OutputWorkspace=norm_ws)
+            Divide(LHSWorkspace=entry,
+                   RHSWorkspace=norm_ws,
+                   OutputWorkspace=ws_name)
+        output_name = self.getPropertyValue('OutputWorkspace')
+        GroupWorkspaces(InputWorkspaces=tmp_names, Outputworkspace=output_name)
+        DeleteWorkspaces(WorkspaceList=[relative_norm_ws, norm_ws])
+        return output_name
+
+    def _set_units(self, ws, nMeasurements):
+        output_unit = self.getPropertyValue('OutputUnits')
+        unit_symbol = 'barn / sr / formula unit'
+        unit = r'd$\sigma$/d$\Omega$'
+        self._set_as_distribution(ws)
+        if output_unit == 'TwoTheta':
+            if mtd[ws].getNumberOfEntries()/nMeasurements > 1 and self.getPropertyValue('OutputTreatment') == 'Sum':
+                self._merge_polarisations(ws)
+                ConvertAxisByFormula(InputWorkspace=ws, OutputWorkspace=ws, Axis='X', Formula='-x')
+            else:
+                ConvertSpectrumAxis(InputWorkspace=ws, OutputWorkspace=ws, Target='SignedTheta', OrderAxis=False)
+                ConvertAxisByFormula(InputWorkspace=ws, OutputWorkspace=ws, Axis='Y', Formula='-y')
+                Transpose(InputWorkspace=ws, OutputWorkspace=ws)
+        elif output_unit == 'Q':
+            if mtd[ws].getNumberOfEntries()/nMeasurements > 1 and self.getPropertyValue('OutputTreatment') == 'Sum':
+                self._merge_polarisations(ws)
+                wavelength = mtd[ws][0].getRun().getLogData('monochromator.wavelength').value # in Angstrom
+                # flips axis sign and converts detector 2theta to momentum exchange
+                formula = '4*pi*sin(-0.5*pi*x/180.0)/{}'.format(wavelength)
```

Comment:
You should use ConvertUnits here, instead of duplicating Q calculation formula. If needed, you can flip the sign before with ConvertAxisByFormula or ScaleX with multiply by -1.

Reply:
The reason that the unit conversion there is done by `ConvertAxisByFormula` instead of `ConvertUnits` is precisely because it does not work there, as the algorithm does not recognize the unit assigned to the axis after `SumOverlappingTubes` is run on the input workspace.

Reply:
Are you referring to x-axis? What if you set the x-axis unit to be scattering angle and then try convert units? If it still doesn't work, then it's fine like this.

Reply:
Yes, the offending axis is the X-axis. There, the unit is already 'Scattering angle', with symbol 'degrees', and unit ID of 'degrees'. The `ConvertUnits` algorithm fails during workspace validation, where the claim is that the workspace has no units, presumably the Y-unit set by `SumOverlappingTubes` is not recognized.